From 3d273731325ed8b93578654406c9ea6080f8ef89 Mon Sep 17 00:00:00 2001
From: Don <don.j.olmstead@gmail.com>
Date: Wed, 29 Apr 2020 13:31:15 -0700
Subject: [PATCH 2/2] Add __has_declspec_attribute

Clang contains __has_declspec_attribute for MSVC compatibility. This can be
used to determine whether __declspec is available.
---
 zconf-ng.h.in | 51 +++++++++++++++++++++++++++++----------------------
 zconf.h.in    | 51 +++++++++++++++++++++++++++++----------------------
 zutil.h       |  2 +-
 3 files changed, 59 insertions(+), 45 deletions(-)

diff --git a/zconf-ng.h.in b/zconf-ng.h.in
index 49075b7..a469a1c 100644
--- a/zconf-ng.h.in
+++ b/zconf-ng.h.in
@@ -23,6 +23,13 @@
 #  endif
 #endif
 
+/* Clang macro for detecting declspec support
+ * https://clang.llvm.org/docs/LanguageExtensions.html#has-declspec-attribute
+ */
+#ifndef __has_declspec_attribute
+#  define __has_declspec_attribute(x) 0
+#endif
+
 /* Maximum value for memLevel in deflateInit2 */
 #ifndef MAX_MEM_LEVEL
 #  define MAX_MEM_LEVEL 9
@@ -53,29 +60,28 @@
                         /* Type declarations */
 
 
-#if defined(WINDOWS) || defined(WIN32)
-   /* If building or using zlib as a DLL, define ZLIB_DLL.
-    * This is not mandatory, but it offers a little performance increase.
-    */
-#  ifdef ZLIB_DLL
-#    if defined(WIN32) && (!defined(__BORLANDC__) || (__BORLANDC__ >= 0x500))
-#      ifdef ZLIB_INTERNAL
-#        define ZEXTERN extern __declspec(dllexport)
-#      else
-#        define ZEXTERN extern __declspec(dllimport)
-#      endif
+/* If building or using zlib as a DLL, define ZLIB_DLL.
+ * This is not mandatory, but it offers a little performance increase.
+ */
+#ifdef ZLIB_DLL
+#  if (defined(WIN32) && (!defined(__BORLANDC__) || (__BORLANDC__ >= 0x500))) || (__has_declspec_attribute(dllexport) && __has_declspec_attribute(dllimport))
+#    ifdef ZLIB_INTERNAL
+#      define ZEXTERN extern __declspec(dllexport)
+#    else
+#      define ZEXTERN extern __declspec(dllimport)
 #    endif
-#  endif  /* ZLIB_DLL */
-   /* If building or using zlib with the WINAPI/WINAPIV calling convention,
-    * define ZLIB_WINAPI.
-    * Caution: the standard ZLIB1.DLL is NOT compiled using ZLIB_WINAPI.
-    */
-#  ifdef ZLIB_WINAPI
-#    include <windows.h>
-     /* No need for _export, use ZLIB.DEF instead. */
-     /* For complete Windows compatibility, use WINAPI, not __stdcall. */
-#    define ZEXPORT WINAPI
-#    define ZEXPORTVA WINAPIV
+#  endif
+#endif  /* ZLIB_DLL */
+
+/* If building or using zlib with the WINAPI/WINAPIV calling convention,
+ * define ZLIB_WINAPI.
+ * Caution: the standard ZLIB1.DLL is NOT compiled using ZLIB_WINAPI.
+ */
+#if defined(WINDOWS) || defined(WIN32) && defined(ZLIB_WINAPI)
+#  include <windows.h>
+   /* No need for _export, use ZLIB.DEF instead. */
+   /* For complete Windows compatibility, use WINAPI, not __stdcall. */
+#  define ZEXPORT WINAPI
+#  define ZEXPORTVA WINAPIV
-#  endif
 #endif
 
diff --git a/zconf.h.in b/zconf.h.in
index f890fcd..1489477 100644
--- a/zconf.h.in
+++ b/zconf.h.in
@@ -23,6 +23,13 @@
 #  endif
 #endif
 
+/* Clang macro for detecting declspec support
+ * https://clang.llvm.org/docs/LanguageExtensions.html#has-declspec-attribute
+ */
+#ifndef __has_declspec_attribute
+#  define __has_declspec_attribute(x) 0
+#endif
+
 /* Maximum value for memLevel in deflateInit2 */
 #ifndef MAX_MEM_LEVEL
 #  define MAX_MEM_LEVEL 9
@@ -57,29 +64,28 @@
 #  define OF(args)  args
 #endif
 
-#if defined(WINDOWS) || defined(WIN32)
-   /* If building or using zlib as a DLL, define ZLIB_DLL.
-    * This is not mandatory, but it offers a little performance increase.
-    */
-#  ifdef ZLIB_DLL
-#    if defined(WIN32) && (!defined(__BORLANDC__) || (__BORLANDC__ >= 0x500))
-#      ifdef ZLIB_INTERNAL
-#        define ZEXTERN extern __declspec(dllexport)
-#      else
-#        define ZEXTERN extern __declspec(dllimport)
-#      endif
+/* If building or using zlib as a DLL, define ZLIB_DLL.
+ * This is not mandatory, but it offers a little performance increase.
+ */
+#ifdef ZLIB_DLL
+#  if (defined(WIN32) && (!defined(__BORLANDC__) || (__BORLANDC__ >= 0x500))) || (__has_declspec_attribute(dllexport) && __has_declspec_attribute(dllimport))
+#    ifdef ZLIB_INTERNAL
+#      define ZEXTERN extern __declspec(dllexport)
+#    else
+#      define ZEXTERN extern __declspec(dllimport)
 #    endif
-#  endif  /* ZLIB_DLL */
-   /* If building or using zlib with the WINAPI/WINAPIV calling convention,
-    * define ZLIB_WINAPI.
-    * Caution: the standard ZLIB1.DLL is NOT compiled using ZLIB_WINAPI.
-    */
-#  ifdef ZLIB_WINAPI
-#    include <windows.h>
-     /* No need for _export, use ZLIB.DEF instead. */
-     /* For complete Windows compatibility, use WINAPI, not __stdcall. */
-#    define ZEXPORT WINAPI
-#    define ZEXPORTVA WINAPIV
+#  endif
+#endif  /* ZLIB_DLL */
+
+/* If building or using zlib with the WINAPI/WINAPIV calling convention,
+ * define ZLIB_WINAPI.
+ * Caution: the standard ZLIB1.DLL is NOT compiled using ZLIB_WINAPI.
+ */
+#if defined(WINDOWS) || defined(WIN32) && defined(ZLIB_WINAPI)
+#  include <windows.h>
+   /* No need for _export, use ZLIB.DEF instead. */
+   /* For complete Windows compatibility, use WINAPI, not __stdcall. */
+#  define ZEXPORT WINAPI
+#  define ZEXPORTVA WINAPIV
-#  endif
 #endif
 
diff --git a/zutil.h b/zutil.h
index 8e9d056..4a456fb 100644
--- a/zutil.h
+++ b/zutil.h
@@ -236,7 +236,7 @@ void ZLIB_INTERNAL   zng_cfree(void *opaque, void *ptr);
 #  define PREFETCH_RW(addr)     addr
 #endif /* (un)likely */
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || __has_declspec_attribute(align)
 #  define ALIGNED_(x) __declspec(align(x))
 #else
 #  if defined(__GNUC__)
-- 
2.25.0.windows.1

